### DISCUSSION TITLE
tutorials/scripting/gdscript/gdscript_styleguide

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#156] [User-contributed notes]
# tutorials/scripting/gdscript/gdscript_styleguide

This style guide lists conventions to write elegant GDScript. The goal is to encourage writing clean, readable code and promote consistency across projects, discussions, and tutorials. Hopefully, t...

https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html

**Comment by shenkes** [comment]
> Code order¶
This first section focuses on code order. For formatting, see Formatting. For naming conventions, see Naming conventions.

FWIW, this section doesn't seem to be the first section anymore.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/9992

**Comment by comminux** [comment]
By the way, where is it better to place network functions (with the @rpc annotation)?

**Comment by stann-co** [comment]
is there any way  to auto format with these conventions. And or make warnings when they aren't followed?

**Reply by Calinou** [reply]
Not yet, see https://github.com/godotengine/godot-proposals/issues/3630.

In the meantime, you can use third-party code formatters such as https://github.com/Scony/godot-gdscript-toolkit, or press <kbd>Ctrl + Shift + I</kbd> with https://github.com/godotengine/godot-vscode-plugin in VS Code.

**Comment by Lexyth** [comment]
> Alternatively, you can use the as keyword to cast the return type, and that type will be used to infer the type of the var.
> [...]
> This option is also considered more type-safe than the first.

Please clarify that "[the] first [option]" refers to inferring the type without a type hint or using a constructor and NOT to the static typing without the `as` keyword, which would be the *first* of the two *static typing* options. 
You could include both options ("These options are...") or just directly mention it ("...than inferring the type without type hints or using a constructor.").

As it stands, this confused even me despite already knowing the difference. A newcomer would just start to use `as` instead of type hints and struggle when perfectly valid variables suddenly turn `null`.

For those interested, the page the hyperlink on 'type-safe' leads to explains why the `as` keyword can be less type-safe:
> The as keyword silently casts the variable to null in case of a type mismatch at runtime, without an error/warning. While this may be convenient in some cases, it can also lead to bugs. Use the as keyword only if this behavior is intended.

**Reply by Lexyth** [reply]
Seems that I confounded null-safety with type-safety here, so the original statement wasn't wrong, but it left out the point about the `as` keyword being less null-safe.

**Comment by UkonnRa** [comment]
about gdscript linting and formatting, is there any official cli solution? I want to run lint & format check in CICD for a Godot repo

**Reply by Calinou** [reply]
There is no official CLI for that yet, but you can use https://github.com/Scony/godot-gdscript-toolkit.

**Reply by docwhat** [reply]
There is also GDQuest's [`gdscript-formatter`](https://github.com/GDQuest/GDScript-formatter), which is dedicated to following the style guidelines.

**Comment by OS-of-S** [comment]
> Prefer the plain English versions of boolean operators, as they are the most accessible

What does it mean? It mean that it's just more understandable and 'easier' to print via keyboard, or does it affects the performance? Why is it even added into an engine if it mentioned as "unrecommened" everywhere? I know that it's a C++ syntax, but why to mention it as "unrecommended" anyway?

PS: this documentation parts is kinda hard to read becouse of all those moments...

**Reply by Lexyth** [reply]
It means you should use `and` instead of `&&` whenever possible, which in turn means *always* unless you are told to use "&&" by someone who has a say in the matter (e.g.: boss/collaborator). 

It's easier to read `and` as "and" than having to translate arbitrary symbols.

There should never be a difference in performance between any keyword aliases in any language. They should always be treated the exact same way by any compiler or interpreter.

It was added because it's the standard symbol meaning "and" in most programming languages and many people are used to it, while `and` was added as the modern alias. That aside, as mentioned, you could be required to use "&&", e.g.: for consistency with other languages used by your team.

It's not recommended because its meaning is less clear than `and` and the symbols may not be readily accessible on all keyboard layouts. The vertical bar (`|`) for example, is situated in an rather inconvenient location on my keyboard.

**Reply by OS-of-S** [reply]
>It's easier to read and as "and" than having to translate arbitrary symbols.

No, it's not. For a not native english-speaker quick glance at the program turns all this words into mess. It's much easer to translate arbitrary symbols becouse it's half-intuitive, short (wich makes whole expression a little shorter) and have different shapes enough to recognize it fast. They also differ significantly from the variable names that so often appear in logical expressions too.
I agree with the same-style-using-for-a-whole-project, but as it doesn't affects the performance I think the documentation shouldn't confuse user in a such way.

**Reply by Lexyth** [reply]
Yes, it is; at least for the vast majority. 
But both styles are supported, so you can use the one that suits you more.

I'm not a native English-speaker either, but neither am I native at arbitrary symbols.
And contrary to those symbols, English has established itself as the closest thing to a universal language we have, so most anyone has had some contact with these words.

Arbitrary symbols like `!`, `^`, `|`, and `~`, on the other hand, **only** have these meanings in programming languages that support them. They are the very opposite of intuitive; you can not guess at their meaning without having learned them in the very context they are used in. 

For example, the symbol "!" conveys "importance" and "urgency" in most any language that uses punctuation, and not the concept of negation that `!` is used for in programming languages. Or, even if not a logical operator, `:` in a certain other programming language means "in". Neither are used like that in any other context I know of.

Words like `not`, `or`, or `and` are much more intuitive for the vast, vast majority, since they are using these outside of the specific domain of programming languages. 

As for them being shorter, that doesn't equate to better. You can at most save 6 characters per line while sticking to proper style conventions. That is pointless and can not affect readability. The improper naming required to save more characters would be the much worse offense regarding readability. 
To clarify, I'm referring to these specifically: 
- identifiers should be short but descriptive (~5-15 chars), 
- lines should be no longer than 80 characters, 
- and you should refactor complex expressions. 
After proper indentation, you've only got space for 3-4 variables, so at most 3 times using `!` instead of `not`.

At least in my opinion, the words differ just as much in shape as the symbols: `&&`, `||`, `!` vs `and`, `or`, `not`

Their similarity to variable names and that they appear between other variables does make these keywords harder to identify, but they're colored in most any code editor, so that shouldn't really be a problem.

The language and documentation are targeted at the majority of people. For the others, there is no to very little difference between having to memorize one set of character sequences over the other. 
If anything, they're trying to do the very opposite of confusing anyone. That's why they explicitly state which one you *should* use.

Besides, arbitrary symbols like `&&` have always been outliers among keywords like `for`, `while`, `class`, and so on, and do not have the justification of being common mathematical symbols like `=`, `+`, or `<`.

I understand that certain design choices can be frustrating, but they have good reasons for why they do certain things (usually for compatibility with the majority of users or systems). 
And in this case, it only affects the documentation and not a feature of the editor or language itself that would prevent you from doing something. 
You are free to choose which version you use in your own code.

**Comment by Gatreh** [comment]
It's missing @icon, which has to come under tool (if it's also a tool).
Also it has a different order to the auto-generated documentation that sorts it as 
Properties (var)
Signals
Enumerations
Constants
Regardless of actual order in the script, which is slightly confusing. We should either update the documentation to reflect the ordering of the style guide or update the style guide to reflect the way documentation is sorted.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10504

**Reply by Lexyth** [reply]
I'd prefer the former, since changing the style guide would make projects inconsistent and might even break things that rely on it, like formatters.
However, I'd also argue that the documentation's order makes sense in its context. You'd usually care more about properties and methods than signals. That is why signals are the first in a script; they are kept further from the "core" concepts. You can't put them below properties and methods in scripts either, because signals still belong closer to data (properties, which are at the top), conceptually, than actions (methods, which come then). 
With documentation, signals are still in a weird spot, but can be defined below methods, while staying close enough to properties, due to methods showing only their declarations without definitions or descriptions.
It's inconsistent, but only because the docs already show properties and methods in a manner that is inconsistent with scripts - all "bodies"(descriptions/definitions) come after all "headers"(declarations).

**Comment by ArtCrafter2999** [comment]
Is there a naming convention for node names?

**Reply by Gatreh** [reply]
Yes, it's PascalCase

**Comment by inputrandomness** [comment]
The section "Methods and static functions" does not mention static function at all. I take it the style guide does not prescribe where static functions should appear? Perhaps this is because they are newer and it has not been decided?

**Reply by Lexyth** [reply]
Definitions should be logically grouped and ordered.

As most static variables and methods belong to the whole class itself, rather than individual instances, I've taken on declaring static variables directly after constants, before the instance variables.
Static methods, *usually*, come just before instance methods; i.e. before `_init`, if present.
That way you separate the behavior of the class itself from its instances'.

Of course, exceptions can be made.
Sometimes, a method or variable is strongly related or directly belongs to its static counterpart, or vice versa.
In those cases, you could group the static version alongside the instance one, but never the other way around; i.e.: no instance method in the static methods section.
E.g.: when the same action can affect all instances in addition to a single one, like blocking a single Channel instance (`first_channel.block()`) vs all Channels (`Channel.blockAll()`).

**Comment by SolsticeMage** [comment]
Regarding [multiline statements](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#format-multiline-statements-for-readability) is it appropriate to format it like this?

```gdscript
var direction := (
          transform.basis * Vector3(input_dir.x, 0, input_dir.y)
).normalized()
```

or like this?


```gdscript
var direction := (
          (transform.basis * Vector3(input_dir.x, 0, input_dir.y))
          .normalized()
)
```

In the first .normalized() seems to add clutter, but in the latter I'm adding extra parentheses. I think I favor the latter, but I wanted to see if there's anything against that approach.

**Reply by SolsticeMage** [reply]
or if I might use this example. I still struggle to know what to do here:

```gdscript
	var input_dir := Input.get_vector("left", "right", "up", "down")
	var direction := (
			(
					transform.basis *
					Vector3(input_dir.x, 0, input_dir.y)
					.rotated(Vector3.UP, camera_arm.rotation.y)
			)
			.normalized()
	)
```

**Reply by Lexyth** [reply]
None are readable, nor are they necessary. 
Multiline statements should do one thing over multiple lines, combining them into a single statement to logically group them.
What you got are multiple things over multiple lines, so you might as well use multiple statements.
For your case, the "correct" approach (as far as conventions and best practices can be considered correct) would be to refactor those into separate steps.
```gdscript
var direction: Vector3 = transform.basis * Vector3(input_dir.x, 0, input_dir.y)
direction = direction.normalized()
```
```gdscript
var input_dir := Input.get_vector("left", "right", "up", "down")
var direction := transform.basis * Vector3(input_dir.x, 0, input_dir.y)
direction = direction.rotated(Vector3.UP, camera_arm.rotation.y)
direction = direction.normalized()
```
You might even want to refactor those steps into their own variables (or functions) to have the code document itself more clearly, as `direction` isn't really a direction until it is normalized.

But those are just suggestions and you can very much get away with cramming it all into a single statement for the more negligible parts.

Edit: If you take another look at the page you linked, you'll find that they all repeat a very similar expression over the multiple lines of the statement.

**Reply by SolsticeMage** [reply]
Thanks for the insight. I've modified my code based on your suggestions. I prefer to avoid creating extra variables in the name of readability. Thought I'd leave my code here in case it helps a kindred spirit.

```py
func process_move_input(delta: float) -> void:
	var input_dir := Input.get_vector("left", "right", "up", "down")
	
	# Moves relative to the camera arm rotation
	var direction := get_relative_input(input_dir, camera_arm.rotation.y)

	# ...
```

```py
# Takes 2D input vector and returns 3D vector direction relative to camera
func get_relative_input(input: Vector2, camera_rotation, axis := Vector3.UP) -> Vector3:
	var rel := Vector3(input.x, 0, input.y)		# Convert 2D input to 3D vector
	rel = rel.rotated(axis, camera_rotation)	# Rotate relative to angle in radians
	rel *= transform.basis						# Scale with transform basis
	return rel.normalized()						# Scale back to 1 unit

```

**Reply by Lexyth** [reply]
Glad I could help.
Funnily enough, in the name of readability, I tend to always create a new variable rather than reassigning values, as usually changing a value means it's not what the variable's name implies anymore; obvious exceptions aside.
In the same line, I tend to create helper functions for anything that represents a single logical group and is more than three lines, similar to your code there.
But that is very much personal preference. As long as it's readable, whether that's achieved through descriptive identifiers or comments, that's all that matters.

**Comment by DexterFstone** [comment]
What is the order of `@abstract` methods? Where do we put them?

**Reply by Lexyth** [reply]
Right after the built-in virtual methods and before the overridden custom methods. 

They affect the entire class, forcing it to become abstract, too, so they get a prominent spot. 
And they're part of the public API, so, again, prominent. 
And they're the "source" of the overridden custom methods, so they should be close to them, too; "close" meaning logically in the general class structure, not literally, since you can't have an abstract method and its implementation within the same file.

**Comment by XandaPanda42** [comment]
Should the convention for signal callback functions to called "`_on_node_name_signal_name()`" be listed here as well? 

It's a fairly common convention, in my opinion, and it's automatic when the editor attempts to create a new function when connecting a Signal **from another node** (in the Node tab on the right hand side).

Example: 
```GDScript
class_name Humanoid

signal passed_out
```

would be automatically set as 

```GDScript
func _on_humanoid_passed_out() -> void:
    pass
```

The editor does it by default, it's mentioned twice in the docs tutorials, but I just realized that I can't find any built-in Nodes that actually do this. (`Node` has `_enter_tree()`, and not `_on_node_enter_tree()`, etc)

**Reply by Lexyth** [reply]
No, because it's not a convention. 
It's just commonly used and the only option for the editor, since it can't infer what the signal emission represents. 

Proper conventions suggest binding methods which represent actual behavior instead. 

So, in your case, it would be something along the lines of: `humanoid.passed_out.connect(fall_over)`

That could then be extended, if additional actions should take place on a "passing out" event, by either additionally connecting other methods that represent proper behavior to the signal, if those actions should stay separate, or by grouping them within a single method/function that still should represent an individual behavior instead of a reaction to an event; in this case, `pass_out`.

**Comment by ghost** [comment]
What should be done about long strings? I want to think that they should be wrapped, and have two indents added to clarify a continuation line, but that introduces newlines and tabs in the string. Newlines can be removed by adding a \ at the end of lines, but what about the tabs? Should string continuations just not use tabs, or is there some other way to handle this?

**Reply by Lexyth** [reply]
You have several options, including joining a String Array with `String.join` or using Multiline-Strings with `String.dedent` in combination with `String.trim_prefix` and `String.trim_suffix`, but I'd prefer simple concatenation:
`String.join`:
```gdscript
var my_long_string: String = " ".join([ # NOTE: joining on a SPACE character in this case, but an empty String can make more sense, depending on context.
    "This",
    "is",
    "a",
    "long",
    "String."
])
```
`String.dedent` with `String.trim_prefix` and `String.trim_suffix`:
```gdscript
# NOTE: String.dedent "pushes" the whole block uniformly, so you can still have individual lines with "extra" indentation or starting with spaces.
# Also, pay special attention to trailing white spaces or tabs, since they're not identifiable without special fonts or the related Editor settings (Draw Tabs and Draw Spaces) activated.
var my_long_string: String = """
    This 
    is 
    a 
    long 
    String.
""".dedent().trim_prefix("\n").trim_suffix("\n")
```
Concatenation (+):
```gdscript
var my_long_string: String = ( # NOTE: parentheses are not optional; indentation is.
    "This "
    + "is "
    + "a "
    + "long "
    + "String."
)
```

**Reply by ghost** [reply]
What do multi-line strings (i.e. """String""") do? I've seen that regular strings can also contain multiple lines, so what's the point of multi-line ones?

**Reply by Lexyth** [reply]
There are pretty few differences.

For one, intent. Multiline string literals explicitly represent a block of text, spanning multiple lines.

Secondly, they allow the use of up to two consecutive instances of the same type of quote as used to delimit the literal itself:
```gdscript
var my_string: String = """You can have two double quotes ("") and two single quotes('') within a multiline string literal, without needing to escape either."""
```

Another thing that I personally think they should—but sadly DON'T—allow, is to have backslashes at the end of a line without the need to escape them. That would be a use case that could set them apart some more, and would make conceptual sense. 
But it's not the case, so the former two seem to be the only aspects that set them apart, as far as I know.

By the way, they should NOT be used as multiline comments, since, while they do get ignored by the parser, just like single-quoted string literals, and thus could be abused in that way, they do not get treated as comments by the editor, which means they do NOT get comment-specific highlighting of the literal itself nor of any tags used within, like "TODO" or "NOTE". And they particularly can NOT be used as docstrings.

**Comment by docwhat** [comment]
Are there any style guidelines for folder/directory names? Should they be PascalCase or snake_case?

**Reply by Calinou** [reply]
I'd say they should be snake_case, like general file names.

**Reply by Lexyth** [reply]
Indeed. There's a section regarding file names in the style guide, too: https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#file-names

In addition to what's mentioned there, most any file created by Godot defaults to being named using the snake case format, so it is going to be the most common one by virtue of that.

**Comment by rethyria** [comment]
Is there any guidance on how to space out @export_category and @export_group? 
Should there be blank lines between them or should all @export vars just be one big list?

**Reply by Lexyth** [reply]
You should use blank lines whenever you want to separate "blocks" of variables that should be separated logically, even if they're of the same kind. These annotations are partly there so you can continue with that while allowing you to combine multiple separate blocks under a common group or category, while maintaining the logical separation of individual blocks.

As for the annotations themselves, you should separate them with a blank line on either side, to make it clear that they're not part of any variable blocks (or groups, for the category annotation) themselves.

**Comment by elementbound** [comment]
```gd
# What type is this? It's not immediately clear to the reader, so it's bad.
var value := complex_function()
```

I see the point, by following the advice, readers will immediately see the variable type because it's right there in the code.

However, I think this also introduces friction for evolving code, because if you change the return type, you have to update every call site. And in editor, it should still be trivial to check the function return type or even the variable type.

I think there is a case for the above example to be at least considered neutral. Thoughts?

**Reply by Lexyth** [reply]
This would be an issue whenever you change to a type that allows the same operations, but produces different results:
```gdscript
func get_my_pet() -> Dog: # later Cat
    return doggy # later kitty

var inferred_value := get_my_pet() # first doggy, later kitty.
var explicit_value: Dog = get_my_pet() # first doggy, later error so we know to do something about this.

print("My dog goes " + inferred_value.speak()) # first "woof", later "meow".
print("My dog goes " + explicit_value.speak()) # first "woof", then, because we'd get an assignment error if it was a Cat, 
# either nothing because we removed this statement or "meow" and we changed the message to "My cat goes ".
```
Either can be very much intended. You might want to suddenly simply switch to printing "meow" everywhere. But if not, you're in deep trouble and have no way to solve it. And since you can't plan for that, you should always use explicit typing. 

You might be fine if a specific call automatically adapted to or just inherently handled "meow", but what if you decide to get a bird or a worm? Will that call still make sense with a worm's speak? "My pet goes .", since worms don't speak and you didn't handle empty strings separately, because Dog guaranteed a non-empty string.

Inference assumes that what it finds there is what is supposed to be there, always, and delegates all problem-catching to the following statements that make use of that variable, which can either complain (good), handle it gracefully (lucky), or not so gracefully (bad). It's a coin toss and the latter can be very difficult to pin-point when it's not part of an obvious print statement, because something changed that you haven't touched in ages.

**Comment by OvercookedBeef** [comment]
I don't see any constants in the full example at the top. From the code order section it would go under the signal and above the @export in the example (since the example doesn't have static variables or enums, those should probably be added too), I think, but it's not clear about the empty lines. For example there appears to be an empty line between regular variables and @onready variables, but not between regular variables and @export variables. Should there be a an empty line with constants next to static variables for example? The code order section shows NO empty line between regular variables and @onready variables which doesn't line up with the example.

**Reply by Lexyth** [reply]
The code order section shows no empty lines between anything other than to split the specifications for class, state, and execution (although it lumps inner classes in with that, likely because it would look weird if separated). I thought the guidelines mentioned how spacing between different sections should be handled, but I can't seem to find it. However, generally, you should separate anything that forms its own logical unit with a single empty line, even within the same kind of variables (e.g., separating motion related constants from ones related to distance), similar to how you're meant to handle this within functions.

**Comment by dougDuarte** [comment]
In the case of private constants, should we also add an underscore? For example: _MY_CONSTANT.

**Reply by dougDuarte** [reply]
Oh, and what about exported variables that I don't want to be accessible outside the class? Should they be written as @export var _my_var?

**Reply by elementbound** [reply]
> Oh, and what about exported variables that I don't want to be accessible outside the class? Should they be written as @export var _my_var?

This is what I tend to do. I like it because in the inspector it still shows up as "My Var", regardless of the underscore at the start.

**Reply by Lexyth** [reply]
> This is what I tend to do. I like it because in the inspector it still shows up as "My Var", regardless of the underscore at the start.

Yup, that's specifically to allow this use case.

As for the private constants, the guide doesn't explicitly mention it, it's not really necessary to protect a read-only variable, and an underscore before a capitalized word looks off, but I'd personally recommend it just for organizational reasons. Filtering out unwanted identifiers keeps the suggestions cleaner. Just make sure to stick to one convention and not change it arbitrarily.

**Comment by ChandelureF-5** [comment]
Dumb question, but which of these would be the visually-correct way of declaring a statically-typed variable(minus the lack of initialization in the examples ^^')?
```gd
var spaced_type: Type

var unspaced_type:Type
```

**Reply by Lexyth** [reply]
The spaced version is, or at least should be, the only one used in the guide; it's easier to distinguish the type from the variable.

**Reply by ChandelureF-5** [reply]
Yeah, that's fair. I just like to do it unspaced for a couple reasons:
1. Familiarity
2. Saving on line length and/or file size

**Reply by Lexyth** [reply]
> 2. Saving on line length and/or file size

Please don't do that.
Familiarity is fine as a reason, though you might want to reconsider that if you plan on contributing to open-source projects or working with a team.
But you should never make any decision based on saving characters. I wrote quite a long text regarding just that somewhere on this page (found it, though slightly different context: https://github.com/godotengine/godot-docs-user-notes/discussions/156#discussioncomment-11763716). 
Readability is a huge factor and saving characters always destroys that for no reason whatsoever.
Code files are inherently small compared to assets, so saving even 5% won't do any good. But you'll lose out on readability, diffing, and grepping and you'll actively fight formatters or you'll have to configure them for this specifically. And you'll have to put in extra work to make sure copy-pasted or generated (by AI, but also any other tool) code is consistent with your format. And you must remember your style to stay consistent, which sounds easier than it is.

Following standards is almost always the right choice, unless you have very, very good reasons not to (i.e., you're forced to use unusual formatting for a necessary tool to work properly).

Of course, it's your choice. This is just advice.
